R/best_case_worst_case.R

```diff
@@ -0,0 +1,114 @@
+#' restore <- options(readr.show_col_types = FALSE)
+#'
+#' companies <- read_csv(toy_emissions_profile_any_companies())
+#' co2 <- read_csv(toy_emissions_profile_products_ecoinvent())
+#' europages_companies <- read_csv(toy_europages_companies())
+#' ecoinvent_activities <- read_csv(toy_ecoinvent_activities())
+#' ecoinvent_europages <- read_csv(toy_ecoinvent_europages())
+#' isic_name <- read_csv(toy_isic_name())
+#' profile_emissions_at_product_level <- profile_emissions(
+#'   companies,
+#'   co2,
+#'   europages_companies = europages_companies,
+#'   ecoinvent_activities = ecoinvent_activities,
+#'   ecoinvent_europages = ecoinvent_europages,
+#'   isic = isic_name
+#' ) |>
+#'   unnest_product()
+#'
+#' result <- best_case_worst_case(profile_emissions_at_product_level)
+#' result
+#'
+#' # Cleanup
+#' options(restore)
+best_case_worst_case <- function(data) {
+  crucial_cols <- c(col_companies_id(), col_ep_product(), col_benchmark(), col_emission_profile())
+  check_matches_col_names(data, crucial_cols)
+
+  data |>
+    mutate(
+      n_distinct_products = n_distinct(.data[[col_ep_product()]], na.rm = TRUE),
+      .by = col_companies_id()
+    ) |>
+    mutate(
+      equal_weight = ifelse(.data[[col_n_distinct_products()]] == 0, NA,
+        1 / .data[[col_n_distinct_products()]]
+      )
+    ) |>
+    lowest_risk_category_per_company_benchmark(.by = c(col_companies_id(), col_benchmark())) |>
+    highest_risk_category_per_company_benchmark(.by = c(col_companies_id(), col_benchmark())) |>
+    mutate(
+      best_risk = assign_risk(
+        .data, col_emission_profile(),
+        col_min_risk_category_per_company_benchmark()
+      )
+    ) |>
+    mutate(
+      worst_risk = assign_risk(
+        .data, col_emission_profile(),
+        col_max_risk_category_per_company_benchmark()
+      )
+    ) |>
+    mutate(
+      count_best_case_products_per_company_benchmark = sum(.data[[col_best_risk()]]),
+      .by = c(col_companies_id(), col_benchmark())
+    ) |>
+    mutate(
+      count_worst_case_products_per_company_benchmark = sum(.data[[col_worst_risk()]]),
+      .by = c(col_companies_id(), col_benchmark())
+    ) |>
+    mutate(
+      best_case = assign_case(
+        .data, col_best_risk(),
+        col_count_best_case_products_per_company_benchmark()
+      )
+    ) |>
+    mutate(
+      worst_case = assign_case(
+        .data, col_worst_risk(),
+        col_count_worst_case_products_per_company_benchmark()
+      )
+    )
+}
+
+lowest_risk_category_per_company_benchmark <- function(data, .by) {
+  risk_order <- c("low", "medium", "high")
+  mutate(data,
+    min_risk_category_per_company_benchmark = risk_first_occurance(.data, risk_order),
+    .by = all_of(.by)
+  )
+}
+
+highest_risk_category_per_company_benchmark <- function(data, .by) {
+  risk_order <- c("high", "medium", "low")
+  mutate(data,
+    max_risk_category_per_company_benchmark = risk_first_occurance(.data, risk_order),
+    .by = all_of(.by)
+  )
+}
```

Comment:
Extracting `risk_order` as an argument would make these two functions identical. But it seems simpler to inline them (i.e. make the caller use  `mutate()` directly rather than calling these interfaces). Usually the benefit of using an interface is that it gives  you the opportunity to define an evocative name that describes what you're doing. But here the column names are already evocative (e.g. `max_risk_category_per_company_benchmark`) and the interfaces mostly duplicate those names and add needless complexity.

Reply:
@maurolepore You are right! I have now removed above functions and passed the `risk_order` parameter as an argument to the `risk_first_occurance` function to apply your understanding. Please review!